app/include/drivers/ext_power.h

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2020 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#pragma once
+
+#include <zephyr/types.h>
+#include <stddef.h>
+#include <device.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @cond INTERNAL_HIDDEN
+ *
+ * Behavior driver API definition and system call entry points.
+ *
+ * (Internal use only.)
+ */
+
+typedef int (*ext_power_enable_t)(struct device *dev);
+typedef int (*ext_power_disable_t)(struct device *dev);
+typedef int (*ext_power_get_t)(struct device *dev, bool *status);
+
+__subsystem struct ext_power_api {
+    ext_power_enable_t enable;
+    ext_power_disable_t disable;
+    ext_power_get_t get;
+};
+/**
+ * @endcond
+ */
+
+/**
+ * @brief Enable the external power output
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 If successful.
+ * @retval Negative errno code if failure.
+ */
+__syscall int ext_power_enable(struct device *dev);
+
+static inline int z_impl_ext_power_enable(struct device *dev) {
+    const struct ext_power_api *api = (const struct ext_power_api *)dev->driver_api;
+
+    if (api->enable == NULL) {
+        return -ENOTSUP;
+    }
+
+    return api->enable(dev);
+}
+
+/**
+ * @brief Disable the external power output
+ * @param dev Pointer to the device structure for the driver instance.
+ *
+ * @retval 0 If successful.
+ * @retval Negative errno code if failure.
+ */
+__syscall int ext_power_disable(struct device *dev);
+
+static inline int z_impl_ext_power_disable(struct device *dev) {
+    const struct ext_power_api *api = (const struct ext_power_api *)dev->driver_api;
+
+    if (api->disable == NULL) {
+        return -ENOTSUP;
+    }
+
+    return api->disable(dev);
+}
+
+/**
+ * @brief Get the current status of the external power output
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param status Status sent to the user
+ *
+ * @retval 0 If successful.
+ * @retval Negative errno code if failure.
+ */
+__syscall int ext_power_get(struct device *dev);
+
+static inline int z_impl_ext_power_get(struct device *dev, bool *status) {
```

Comment:
Hmmm.... Thoughts on instead using negative return integers for error, `0` for off, `1` for on?

Reply:
Yeah I considered that, but should we return '0' if some driver does not implement get (which is most unlikely).
Also I can definitely see the benefits of returning status in the return code, which will simplify application code by calling the api without the need for extra `status` variable.

I observed some zephyr driver api's returning status directly.

Its basically boils down to project preference I think, and its better if it is consistent across the entire project. So its your decision

Reply:
If the specific driver doesn't support get, it should be returning `-ENOTSUP` anyways, so I don't think that's a problem.

I think we should stick to a single int return here, yes. I find it much preferable to bool pointer + return code.

Reply:
> If the specific driver doesn't support get, it should be returning -ENOTSUP anyways, so I don't think that's a problem.

Once the application got used with the boolean return status, like
```
if (get()) {
    do_something();
}
```
returning -ENOTSUP for unsupported apis will result in the above been succeed.

But I am all in for single int return since it application code a lot

Reply:
I think the key here is to have the code do:

```
if (get() > 0) {
  do_something();
}
```

which is still reasonable.